homeassistant/components/mqtt/__init__.py

```diff
@@ -370,18 +389,30 @@ def async_publish_service(call):
         msg_topic = call.data[ATTR_TOPIC]
         payload = call.data.get(ATTR_PAYLOAD)
         payload_template = call.data.get(ATTR_PAYLOAD_TEMPLATE)
+        payload_file_path = call.data.get(ATTR_PAYLOAD_FILE_PATH)
         qos = call.data[ATTR_QOS]
         retain = call.data[ATTR_RETAIN]
-        try:
-            if payload_template is not None:
+        if payload_template is not None:
+            try:
                 payload = \
                     template.Template(payload_template, hass).async_render()
-        except template.jinja2.TemplateError as exc:
-            _LOGGER.error(
-                "Unable to publish to '%s': rendering payload template of "
-                "'%s' failed because %s",
-                msg_topic, payload_template, exc)
-            return
+            except template.jinja2.TemplateError as exc:
+                _LOGGER.error(
+                    "Unable to publish to '%s': rendering payload template of "
+                    "'%s' failed because %s",
+                    msg_topic, payload_template, exc)
+                return
+        elif payload_file_path is not None:
```

Comment:
@pvizeli I think I understood @balloob just fine. What I don't understand is the lack of acknowledgement of the underlying problem and of some suggestion for alternate solutions. To be blunt, the underlying problem is that the base implementation of MQTT in HA was not protocol compliant. Can we please find a solution?

Given the very limited understanding I have of HA, I humbly offer a few thoughts:
It looks to me that there's no way, at the moment, a component could load a file or, for what is worth noting, build a binary chunk of data and publish it on an MQTT topic _as it is_ through the MQTT service. We are forced to put it into a JSON formatted string which, obviously, cannot contain binary data. Am I wrong?

Now, if this is a base, transport component, why offer a service at all? Shouldn't we have a separate component for that, one that manages _any_ kind of payload, as per mqtt specs?

IMHO, this ambiguity is well expressed, involuntary, in [this page](https://home-assistant.io/components/notify.mqtt/). I'm sure there are historical and practical reasons for this situation but, humbly, I dissent: this doesn't look like a good implementation to me.

So, you seem to be saying, I should write a new component that does exactly what the MQTT service does, except that it is not a service and it accepts binary payloads. Is that it?

thanks,
gl